aip/general/0129.md

```diff
@@ -0,0 +1,71 @@
+Imperative clients rely on the result of an RPC call to determine success or failure of an operation. State-driven clients (including declarative clients) are clients that use the state of an existing resource to determine actions that should be taken to bring that resource to a desired state. To determine if a set of updates was successful a state-driven client compares server side resource attributes with client defined values. The comparison feature ensures accuracy of a creation or an update but it requires services to treat the client set fields as read-only and diligently preserve those values.
+
+### Permadiff
+Services that update client-set fields communicate to state-driven clients that the desired change has not taken effect, which results in the state-driven client to continuously call the service with updates. Permadiff is an infinite loop scenario where a client continuously attempts to update a resource field or a set of fields that the API service also modifies.
+
+## Guidance
+
+### Server Side Defaults
+There are instances where a service has to ensure that a value always exists even if the client chooses not to specify one. For example: a client may create a caching service instance without specifying the software version to be used for the service. Such a scenario requires the service to pick a version. The service designer has a number of ways to handle server side defaults.
+
+#### Dynamic Defaults
+When a default value of a field may change at any point in time such a field is considered to have a dynamic default. In general, dynamic fields can be: 
+* generated (UUID)
+* allocated (dynamic IP address)
+* assigned (most recent software package version)
+
+An attribute with a dynamic default ***must*** be expressed as two fields in the API:
+* a mutable field that can be optionally set by the user
+* an immutable `OUTPUT_ONLY` field that records the final value decided on by the API service
+
+Example:
+```proto
+message VirtualMachine {
+  …
+  string ip_address = 4;
+  string provisioned_ip_address = 5 [
+    (google.api.field_behavior) = OUTPUT_ONLY
+  ];
+}
+```
+
+#### Static Defaults
+Service ***may*** choose to designate a value to be an equivalent to an empty value. This type of default is convenient for values that don't change very often since a change to a static default is considered a non-backward-compatible change within the same major API version as specified in [AIP-180](aip-180). To designate a value as a static default a service must add google.api.default annotation to the field definition. Annotating a field with google.api.default specifies machine-readable documentation of the field's behavior, allowing clients to understand default behavior and the meaning of an empty field.
+
+Example:
+```proto
+message VM {
+ string sku = 1 [(google.api.default) = "foo"];
+}
+```
+
+### Mutable fields
+API Services ***must not*** modify mutable fields.
+
+### Semantic Strings Handling
```

Comment:
"Semantic string" was the term used to describe a data type represented in the API using a string: uuids, IPv4 addresses, DNS RFC1123 names, etc. These have specific value validation rules and sometimes are represented in systems in ways that don't preserve the original client-provided string values.

Reply:
Right. I think it's fine as a concept - I'd just prefer to use well-known terms... or if we're *going* to define a new term, make it really clear that we're doing so. (A reader shouldn't read the section and feel bad for not knowing what a term means already, if we've just made it up...)

Reply:
I don't love the term. I was just explaining it. :-)

I'd probably just use something like "data types represented as canonicalized strings".

Reply:
"Field Value Handling"? "Semantic Value vs Marshaled Representation"?

It applies to strings, but isn't limited to them. Not certain what the term should be but should definitely cover all of those cases.

Personally, I think of this as "Value entropy vs Marshaled entropy", which isn't more clear.

But with a uuid for example, the field value has 2^128 distinct values, exactly the same value space as 32 hex characters (16^32). Literal comparison of [0-9a-zA-Z] has more distinct values (22^32) which is why you *must* perform all equality+comparison operations on the value space, not in the representation space.

This is really good to connect to the above concept also. Having a defined default on a field actually means that it _is not a server modified value_ --- they are different representations of the same value.

```
enum ColorEnum {UNSPECIFIED, RED, GREEN, BLUE};
ColorEnum color; 
```
The entropy of color is 4, because the field may be in the states of `{UNSPECIFIED, RED, GREEN, BLUE}`

Or a different angle with string rather than enum mechanically:
```
// optional field, valid colors include: "red", "green", "blue"
string color_name; 
```
The entropy of color is 4, because the field may be in the states of `{empty, "red", "green", "blue"}`. Even if you tolerates "Red" and "RED" the entropy of the value space is still 4.

It's when you declare the default value - you've defined the semantic value of the empty representation.
```
enum ColorEnum {UNSPECIFIED, RED, GREEN, BLUE};
ColorEnum color [(google.api.default) = ColorEnum.RED]; 

// valid colors include: "red", "green", "blue". default is "red".
string color_name [(google.api.default) = "red"); 
```
The entropy is now 3, because (UNSPECIFIED==RED) and (""=="red") _for all intents and purposes_.

So if you send number's json text of 7.000 and get back 7, uuid with "1234abcd" and get back "1234ABCD", send color of `UNSPECIFIED` and get back `RED`, or `""` and get back `"red"` ---- the key concept is that _the server has not modified the value_.

Reply:
Numbers are an interesting case of this, we sometimes see odd behavior in service's decisions about precision and trailing zeros: https://github.com/hashicorp/terraform-provider-google/issues/15166

I'm not sure how to handle for this. Potentially a defined type for duration and we can implement the comparison client-side?

Overall, I changed this section to not be specific to strings as numbers have the same issues